File: source/com/barrybecker4/simulation/habitat/HabitatSimulator.java
/** Copyright by Barry G. Becker, 2000-2015. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.habitat;

import com.barrybecker4.common.concurrency.ThreadUtil;
import com.barrybecker4.common.math.MathUtil;
import com.barrybecker4.simulation.common.ui.Simulator;
import com.barrybecker4.simulation.common.ui.SimulatorApplet;
import com.barrybecker4.simulation.common.ui.SimulatorOptionsDialog;
import com.barrybecker4.simulation.habitat.creatures.Populations;
import com.barrybecker4.simulation.habitat.creatures.SerengetiPopulations;
import com.barrybecker4.simulation.habitat.options.DynamicOptions;
import com.barrybecker4.simulation.habitat.options.HabitatOptionsDialog;
import com.barrybecker4.ui.util.GUIUtil;

import javax.swing.*;
import java.awt.*;

/**
 * Simulates foxes (predators) and rabbits (prey) in the wild.
 *
 * @author Barry Becker
 */
public class HabitatSimulator extends Simulator {

    private Populations populations;
    private DynamicOptions options;
    private JSplitPane splitPane;


    /** Constructor */
    private HabitatSimulator() {

        super("Habitat Simulation");
        setBackground(Color.WHITE);
        populations = new SerengetiPopulations();

        PopulationGraphPanel graphPanel = new PopulationGraphPanel(populations);
        HabitatPanel habitatPanel = new HabitatPanel(populations);

        splitPane = new JSplitPane(JSplitPane.VERTICAL_SPLIT,
                           habitatPanel, graphPanel);


        //Provide minimum sizes for the two components in the split pane
        Dimension minimumSize = new Dimension(100, 50);
        habitatPanel.setMinimumSize(minimumSize);
        graphPanel.setMinimumSize(minimumSize);

        splitPane.setDividerLocation(350);

        this.add(splitPane);
    }

    @Override
    protected void reset() {

        this.setPaused(true);
        // wait till actually paused. Not clean, but oh well.
        ThreadUtil.sleep(500);
        MathUtil.RANDOM.setSeed(1);
        populations.reset();
        options.reset();
        this.setPaused(false);
    }

    @Override
    protected double getInitialTimeStep() {
        return 1.0;
    }

    @Override
    public double timeStep() {

        options.update();
        populations.nextDay();
        return timeStep_;
    }

    public Populations getPopulations() {
        return populations;
    }

    /**
     * Draw the population graph under the hab.
     * @param g java graphics context
     */
    @Override
    public void paint( Graphics g ) {
        splitPane.setSize(getSize());
        splitPane.paint(g);
    }

    @Override
    public JPanel createDynamicControls() {
        options = new DynamicOptions(this);
        return options;
    }

    @Override
    protected SimulatorOptionsDialog createOptionsDialog() {
         return new HabitatOptionsDialog(frame, this);
    }

    public static void main( String[] args ) {
        final HabitatSimulator sim = new HabitatSimulator();

        sim.setPaused(true);
        GUIUtil.showApplet(new SimulatorApplet(sim));
    }
}

File: scala-source/com/barrybecker4/simulation/habitat/HabitatSimulator.scala
// Copyright by Barry G. Becker, 2016-2023. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.habitat

import com.barrybecker4.common.concurrency.ThreadUtil
import com.barrybecker4.math.MathUtil
import com.barrybecker4.simulation.common.ui.Simulator
import com.barrybecker4.simulation.common.ui.SimulatorFrame
import com.barrybecker4.simulation.common.ui.SimulatorOptionsDialog
import com.barrybecker4.simulation.habitat.creatures.populations.Habitat
import com.barrybecker4.simulation.habitat.ui.options.DynamicOptions
import com.barrybecker4.simulation.habitat.creatures.CreatureProcessor
import com.barrybecker4.simulation.habitat.ui.renderers.CreatureRenderer

import javax.swing.*
import java.awt.*
import com.barrybecker4.simulation.habitat.ui.HabitatSplitPanel


/**
  * Simulates foxes (predators) and rabbits (prey) in the wild.
  *
  * @author Barry Becker
  */
object HabitatSimulator {
  def main(args: Array[String]): Unit = {
    val sim = new HabitatSimulator
    sim.setPaused(true)
    val frame = new SimulatorFrame(args.toIndexedSeq, sim)

    frame.pack()
    frame.setLocationRelativeTo(null)
    frame.setVisible(true)
  }
}

class HabitatSimulator extends Simulator("Habitat Simulation") {

  setBackground(Color.WHITE)
  private var iterationsPerFrame = DynamicOptions.INITIAL_ITERATIONS_PER_FRAME
  private var useContinuousIteration = DynamicOptions.DEFAULT_USE_CONTINUOUS_ITERATION
  private var habitat = Habitat.HABITATS(Habitat.DEFAULT_HABITAT_INDEX)
  private var options: DynamicOptions = _
  private var splitPanel = new HabitatSplitPanel(habitat)
  this.add(splitPanel)

  override protected def reset(): Unit = {
    this.setPaused(true)
    // wait till actually paused.
    ThreadUtil.sleep(500)
    MathUtil.RANDOM.setSeed(1)
    habitat.reset()
    options.reset()
    this.setPaused(false)
  }

  override def getPreferredSize: Dimension = new Dimension(1200, 1000)

  override protected def getInitialTimeStep = 0.1

  override def timeStep: Double = {
    if (useContinuousIteration) {
      for (i <- 0 until iterationsPerFrame) {
        options.update()
        habitat.nextDay()
      }
    }
    tStep
  }

  def getHabitat: Habitat = habitat

  def setHabitat(habitat: Habitat): Unit = {
    this.habitat = habitat
    this.remove(splitPanel)
    splitPanel = new HabitatSplitPanel(habitat)
    this.add(splitPanel)
    this.reset()
    ThreadUtil.sleep(100)
    this.setPaused(true)
  }

  def setDebug(isDebug: Boolean): Unit = {
    CreatureProcessor.DEBUG = isDebug
    CreatureRenderer.DEBUG = isDebug
  }

  def setNumPixelsPerXPoint(numPixels: Int): Unit = {
    splitPanel.setNumPixelsPerXPoint(numPixels)
  }

  def setIterationsPerFrame(iterationsPerFrame: Int): Unit = {
    this.iterationsPerFrame = iterationsPerFrame
  }

  def setUseContinuousIteration(useContinuous: Boolean): Unit = {
    this.useContinuousIteration = useContinuous
  }

  def requestNextStep(): Unit = {
      habitat.nextDay()
      splitPanel.repaint()
  }

  /** Draw the population graph under the habitat. */
  override def paint(g: Graphics): Unit = {
    splitPanel.setSize(getSize)
    splitPanel.paint(g)
  }

  override def createDynamicControls: JPanel = {
    options = new DynamicOptions(this)
    options
  }

  override protected def createOptionsDialog = new SimulatorOptionsDialog(frame, this)
}